jetty-websocket/jetty-websocket-client/src/main/java/org/eclipse/jetty/websocket/client/impl/JettyClientUpgradeRequest.java

```diff
@@ -43,16 +45,24 @@
 {
     private final WebSocketClient containerContext;
     private final Object websocketPojo;
-    private final CompletableFuture<Session> futureSession;
     private final DelegatedJettyClientUpgradeRequest handshakeRequest;
+    private final CompletableFuture<Session> futureSession;
+    private final JettyWebSocketFrameHandler frameHandler;
+    private UpgradeResponse upgradeResponse;
 
     public JettyClientUpgradeRequest(WebSocketClient clientContainer, WebSocketCoreClient coreClient, UpgradeRequest request,
                                      URI requestURI, Object websocketPojo)
     {
         super(coreClient, requestURI);
         this.containerContext = clientContainer;
         this.websocketPojo = websocketPojo;
-        this.futureSession = new CompletableFuture<>();
+        CompletableFuture<Session> completableFutureSession = new CompletableFuture<>();
+        this.futureSession = completableFutureSession.thenApply(s->
+        {
+            // Note: we must set the upgrade response before the future session will complete
+            ((WebSocketSession)s).setUpgradeResponse(upgradeResponse);
```

Comment:
I think this is too late to set the upgrade response.  This future is completed after the call of `onOpen` when the session is exposed to application code which could call `getUpgradeResponse`.     So the response needs to be set on the session prior to `onOpen` being called.  I'd also prefer to avoid using  the completablefuture mechanism if possible.

Reply:
Also session listeners are called prior to the future being completed and `onOpen` being called.  So perhaps you could ensure that the upgradeResponse is set by the first listener?